src/Npgsql.GeoJSON/GeoJsonHandler.cs

```diff
@@ -0,0 +1,796 @@
+            {
+                case GeoJSONObjectType.Point:
+                    return ValidateAndGetLength((Point)value, ref lengthCache, parameter);
+
+                case GeoJSONObjectType.LineString:
+                    return ValidateAndGetLength((LineString)value, ref lengthCache, parameter);
+
+                case GeoJSONObjectType.Polygon:
+                    return ValidateAndGetLength((Polygon)value, ref lengthCache, parameter);
+
+                case GeoJSONObjectType.MultiPoint:
+                    return ValidateAndGetLength((MultiPoint)value, ref lengthCache, parameter);
+
+                case GeoJSONObjectType.MultiLineString:
+                    return ValidateAndGetLength((MultiLineString)value, ref lengthCache, parameter);
+
+                case GeoJSONObjectType.MultiPolygon:
+                    return ValidateAndGetLength((MultiPolygon)value, ref lengthCache, parameter);
+
+                case GeoJSONObjectType.GeometryCollection:
+                    return ValidateAndGetLength((GeometryCollection)value, ref lengthCache, parameter);
+
+                default:
+                    throw UnknownPostGisType();
+            }
+        }
+
+        public int ValidateAndGetLength(Point value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
+            => SizeOfHeader + SizeOfPoint(HasZ(value.Coordinates));
+
+        public int ValidateAndGetLength(LineString value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
+        {
+            var coordinates = value.Coordinates;
+            if (NotValid(coordinates, out var hasZ))
+                throw AllOrNoneCoordiantesMustHaveZ(parameter, nameof(LineString));
+
+            var length = SizeOfHeaderWithLength + coordinates.Count * SizeOfPoint(hasZ);
+            if (value.CRS != null)
+                length += sizeof(int);
+
+            return length;
+        }
+
+        public int ValidateAndGetLength(Polygon value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
+        {
+            var lines = value.Coordinates;
+            var length = SizeOfHeaderWithLength + SizeOfLength * lines.Count;
+            if (value.CRS != null)
+                length += sizeof(int);
+
+            var hasZ = false;
+            for (var i = 0; i < lines.Count; ++i)
+            {
+                var coordinates = lines[i].Coordinates;
+                if (NotValid(coordinates, out var lineHasZ))
+                    throw AllOrNoneCoordiantesMustHaveZ(parameter, nameof(Polygon));
+
+                if (hasZ != lineHasZ)
+                {
+                    if (i == 0) hasZ = lineHasZ;
+                    else throw AllOrNoneCoordiantesMustHaveZ(parameter, nameof(LineString));
+                }
+
+                length += coordinates.Count * SizeOfPoint(hasZ);
+            }
+
+            return length;
+        }
+
+        static bool NotValid(ReadOnlyCollection<IPosition> coordinates, out bool hasZ)
```

Comment:
All callers of `NotValid` seems to throw `AllOrNoneCoordiantesMustHaveZ` if it returns true, so why not just throw it from inside the function? You could also rename from the somewhat confusing `NotValid` to, say, `CheckZ`.

Reply:
Will do.

Reply:
The ` ValidateAndGetLength(Polygon, ref NpgsqlLengthCache, NpgsqlParameter)` method throws the exception in two places which are observable. It's more obvious.